File: web-plugin/src/main/java/com/hh/uiperception/webplugin/WebJsonParser.java
package com.hh.uiperception.webplugin;

import com.hh.uiperception.core.semantic.Bounds;
import com.hh.uiperception.core.semantic.SemanticNode;
import com.hh.uiperception.core.semantic.SemanticRole;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * Parses the JSON output from dom-serializer.js into a SemanticNode tree.
 *
 * Data flow: WebView → evaluateJavascript(dom-serializer.js) → JSON string → this parser → SemanticNode tree
 *
 * Expected JSON structure:
 * <pre>{@code
 * {
 *   "url": "https://...",
 *   "title": "Page Title",
 *   "root": {
 *     "role": "screen",
 *     "name": "",
 *     "states": [],
 *     "bounds": [],
 *     "__pr_idx": 0,
 *     "children": [
 *       { "role": "button", "name": "Submit", "states": ["clickable"], "__pr_idx": 1, "bounds": [10,20,100,50], "children": [] }
 *     ]
 *   }
 * }
 * }</pre>
 */
public final class WebJsonParser {

    private WebJsonParser() {}

    /** Entry point: parse the top-level JSON and extract the "root" node. */
    public static SemanticNode parse(String json) {
        if (json == null || json.trim().isEmpty()) {
            return null;
        }
        try {
            JSONObject root = new JSONObject(json);
            JSONObject rootNode = root.optJSONObject("root");
            if (rootNode == null) {
                return null;
            }
            return parseNode(rootNode);
        } catch (JSONException e) {
            throw new IllegalArgumentException("Invalid JSON", e);
        }
    }

    /**
     * Recursively parse a JSON node into a SemanticNode.
     * Maps JSON fields: role → SemanticRole, name, bounds, states, __pr_idx → webElementIdx, children → recursive.
     */
    private static SemanticNode parseNode(JSONObject obj) throws JSONException {
        String roleStr = obj.optString("role", "");
        SemanticRole role = mapRole(roleStr);
        if (role == null) {
            return null;
        }

        String name = obj.optString("name", "");
        Bounds bounds = parseBounds(obj.optJSONArray("bounds"));

        SemanticNode.Builder builder = SemanticNode.builder(role)
                .name(name)
                .bounds(bounds)
                .webElementIdx(obj.optInt("__pr_idx", -1));

        JSONArray states = obj.optJSONArray("states");
        if (states != null) {
            for (int i = 0; i < states.length(); i++) {
                builder.addState(states.getString(i));
            }
        }

        JSONArray children = obj.optJSONArray("children");
        if (children != null) {
            for (int i = 0; i < children.length(); i++) {
                JSONObject child = children.optJSONObject(i);
                if (child != null) {
                    SemanticNode childNode = parseNode(child);
                    if (childNode != null) {
                        builder.addChild(childNode);
                    }
                }
            }
        }

        return builder.build();
    }

    /**
     * Map a role string from JS to SemanticRole enum.
     * Matches against SemanticRole.snapshotName(); unknown roles default to GENERIC.
     */
    private static SemanticRole mapRole(String role) {
        if (role == null || role.isEmpty()) {
            return SemanticRole.GENERIC;
        }
        for (SemanticRole sr : SemanticRole.values()) {
            if (sr.snapshotName().equals(role)) {
                return sr;
            }
        }
        return SemanticRole.GENERIC;
    }

    /** Parse a [x1, y1, x2, y2] JSON array into a Bounds object. */
    private static Bounds parseBounds(JSONArray arr) throws JSONException {
        if (arr == null || arr.length() != 4) {
            return null;
        }
        return new Bounds(arr.getInt(0), arr.getInt(1), arr.getInt(2), arr.getInt(3));
    }
}

File: ios/UIPerception/Capture/WebJsonParser.swift
import Foundation

/// 把 dom-serializer.js 输出的 JSON 解析为 SemanticNode 树。
/// 对应 Android `WebJsonParser.java`。
///
/// 数据流：WKWebView → evaluateJavaScript(dom-serializer.js) → JSON 字符串
///       → 本 parser → SemanticNode 树 → TreeNormalizer → RefAssigner("w") → YAML
enum WebJsonParser {

    /// 解析顶层 JSON，提取 "root" 节点。
    /// 输入是 dom-serializer.js 直接 return 的 JSON 字符串。
    static func parse(_ json: String?) -> SemanticNode? {
        guard let json, !json.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return nil
        }
        guard let data = json.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let rootNode = object["root"] as? [String: Any]
        else { return nil }
        return parseNode(rootNode)
    }

    /// 递归解析 JSON node 为 SemanticNode。
    /// JSON 字段映射：role → SemanticRole、name、bounds、states、__pr_idx → webElementIdx、children → 递归。
    private static func parseNode(_ obj: [String: Any]) -> SemanticNode? {
        guard let roleStr = obj["role"] as? String else { return nil }
        let role = mapRole(roleStr)

        let name = (obj["name"] as? String) ?? ""
        let bounds = parseBounds(obj["bounds"])

        let builder = SemanticNode.builder(role)
            .name(name)
            .bounds(bounds)
            .webElementIdx((obj["__pr_idx"] as? Int) ?? -1)

        if let states = obj["states"] as? [String] {
            for state in states { builder.addState(state) }
        }

        if let children = obj["children"] as? [[String: Any]] {
            for child in children {
                if let childNode = parseNode(child) {
                    builder.addChild(childNode)
                }
            }
        }

        return builder.build()
    }

    /// JSON 字符串 role 映射到 SemanticRole。
    /// 匹配 SemanticRole.snapshotName；未知 role 兜底为 .generic。
    private static func mapRole(_ role: String) -> SemanticRole {
        if role.isEmpty { return .generic }
        return SemanticRole(rawValue: role) ?? .generic
    }

    /// 把 [x1, y1, x2, y2] JSON 数组解析为 Bounds。
    private static func parseBounds(_ any: Any?) -> Bounds? {
        guard let arr = any as? [Int], arr.count == 4 else { return nil }
        return Bounds(left: arr[0], top: arr[1], right: arr[2], bottom: arr[3])
    }
}
